File: BennuLib/IO/MapSpriteEncoder.vb
Option Infer On

Namespace Bennu.IO
    Public Class MapSpriteEncoder
        Inherits NativeEncoder(Of Sprite)

        Protected Overrides ReadOnly Property Version As Byte = &H0

        Protected Overrides Sub WriteNativeFormat(sprite As Sprite, writer As NativeFormatWriter)
            writer.Write(Convert.ToUInt16(sprite.Width))
            writer.Write(Convert.ToUInt16(sprite.Height))
            writer.Write(Convert.ToUInt32(sprite.Id.GetValueOrDefault))

            If Not sprite.Palette Is Nothing Then
                writer.Write(sprite.Palette)
                writer.WriteReservedPaletteGammaSection()
            End If

            Dim ids = sprite.PivotPoints.Select(Function(p) p.Id)

            writer.Write(Convert.ToUInt16(If(ids.Count > 0, ids.Max, 0)))
            writer.Write(sprite.PivotPoints)
            'TODO
            writer.Write(DirectCast(sprite.Pixels, Int32PixelARGB()))
        End Sub

        Protected Overrides Function GetFileId(obj As Sprite) As String

            If TypeOf (obj.Pixels(0)) Is IndexedPixel Then
                Return "map"
            ElseIf TypeOf (obj.Pixels(0)) Is Int16Pixel565 Then
                Return "m16"
            ElseIf TypeOf (obj.Pixels(0)) Is Int32PixelARGB Then
                Return "m32"
            Else
                Throw New ArgumentException
            End If

        End Function

    End Class
End Namespace

File: BennuLib/BennuLib.Core/IO/MapSpriteEncoder.cs
using System;
using System.Data;
using System.Linq;

namespace Bennu.IO
{
    public class MapSpriteEncoder : NativeEncoder<Sprite>
    {
        private const int version = 0x00;

        protected override byte GetLastHeaderByte ( Sprite sprite ) => version;

        protected override void WriteNativeFormatBody ( Sprite sprite, NativeFormatWriter writer )
        {
            writer.Write ( Convert.ToUInt16 ( sprite.Width ) );
            writer.Write ( Convert.ToUInt16 ( sprite.Height ) );
            writer.Write ( Convert.ToUInt32 ( sprite.Id.GetValueOrDefault () ) );

            if ( ( sprite.Palette != null ) )
            {
                writer.Write ( sprite.Palette );
                writer.WriteReservedPaletteGammaSection ();
            }

            var ids = sprite.PivotPoints.Select ( p => p.Id );

            writer.Write ( Convert.ToUInt16 ( ids.Count () > 0 ? ids.Max () : 0 ) );
            writer.Write ( sprite.PivotPoints );
            writer.Write ( sprite.PixelData );
        }

        protected override string GetFileMagic ( Sprite sprite )
        {

            switch ( sprite.Depth )
            {
                case 1:
                    return "m01";
                case 8:
                    return "map";
                case 16:
                    return "m16";
                case 32:
                    return "m32";
                default:
                    throw new ArgumentException ();
            }
        }

    }
}
